### PR TITLE
Add article on atomic design in SPAs

### PR DESCRIPTION
**Pull request by Keats**

### CODE REVIEW

File: content/blog/atomic-design-for-spas.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+date = 2018-10-25
++++
+
+This post will outline how I architect visual part of the SPAs I work on. I used
+to put every components in a `components` folder, which obviously doesn't scale and make
+it hard finding a specific component. After that, I grouped at least the top level routing
+components which made things a bit tidier but was still very unsatisfying.
+Looking around for various ways to organise design elements, I found [Atomic Design](http://bradfrost.com/blog/post/atomic-web-design/).
+I am not going to dive deep into explaning what is Atomatic Design as the linked article does a very good job
+so I would recommend taking 2 minutes to read it beforehand.
```
Comment (AdrienLemaire) [inline content/blog/atomic-design-for-spas.md:13]: looks like it'll take more than 2 min to read though

File: content/blog/atomic-design-for-spas.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+What this means is that every component folder looks very similar to something like:
+
+```bash
+input
+├── index.tsx
+└── _style.scss
+```
+
+You can replace `tsx` by `js` if you are not using [TypeScript](https://www.typescriptlang.org/).
```
Comment (AdrienLemaire) [inline content/blog/atomic-design-for-spas.md:36]: you can replace tsx by js even if you are using typescript. babel/webpack ftw

File: content/blog/atomic-design-for-spas.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+As you can see, the folder also contains a `_style.scss`. There might be some use-cases for CSS in JS but I still haven't seen a compelling one
```
Comment (AdrienLemaire) [inline content/blog/atomic-design-for-spas.md]: Well the compelling arguments are http://cssinjs.org/benefits/?v=v9.8.7

File: content/blog/atomic-design-for-spas.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+If I am using [GraphQL](https://graphql.org/) with [Apollo](https://www.apollographql.com/), I will also have a `queries.tsx` in there with the queries
+used by the component as well as the types:
+
+
+```tsx
+import gql from "graphql-tag";
+import {Query} from "react-apollo";
+
+import {IOrganization} from "../../../types";
```
Comment (AdrienLemaire) [inline content/blog/atomic-design-for-spas.md:60]: babel-plugin-root-import: `import {IOrganization} from "~/types";`

File: content/blog/atomic-design-for-spas.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+## Atomic Design
+
+If you have read the article introducing Atomic Design, you can probably guess that your `components` folder is going to look like that:
+
+```bash
+.
+├── atoms
+├── molecules
+├── organisms
```
Comment (AdrienLemaire) [inline content/blog/atomic-design-for-spas.md:100]: mouarf. atoms/molecules/organisms are a nice analogy, but from there to name your production code this way ....  xD

File: content/blog/atomic-design-for-spas.md [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+└── pages
+```
+
+Let's look at them in the same order.
+
+### Atoms
+
+Atoms are independent components that do not have any dependencies on other components and that exist in isolation: they
+probably don't do anything on their own, are mostly composed of `props` and any action they do is coming from `props`.
+In terms of Redux/MobX, all of those components should be dumb components or presentation-only components.
```
Comment (AdrienLemaire) [inline content/blog/atomic-design-for-spas.md]: dumb = pure

File: content/blog/atomic-design-for-spas.md [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+### Molecules
+
+By themselves, atoms are not very useful and need to be combined to make something useable: `input` + `button` giving you a `form` for example.
+That is what molecules are: a combination of 2 or more `atoms` or, in rarer cases, a smart wrapper around an atom.
+Therefore molecules should be dumb components when possible but it isn't required.
+
+Here are some of the molecules from the same project:
+
+- `form` (`input` + `label`): using MobX for state but dumb otherwise
```
Comment (AdrienLemaire) [inline content/blog/atomic-design-for-spas.md:139]: HOC or container. pure as well

File: content/blog/atomic-design-for-spas.md [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+- `login`
+- `home`
+- `user-settings`
+- `verify-email`
+
+
+## Conclusion
+
+While some of the rules I introduced seem inflexible, you can bend them so they make sense to you.
+For example, I put my `overlay` component in the atoms folder despite it using the `portal` atom simply because I kept looking for it in the `atoms` folder
```
Comment (AdrienLemaire) [inline content/blog/atomic-design-for-spas.md]: :rofl:

File: content/blog/atomic-design-for-spas.md (hunk 1 above)
Comment (Keats) [reply content/blog/atomic-design-for-spas.md:13]: it seems long but there are lots of image so it is pretty fast to read

File: content/blog/atomic-design-for-spas.md (hunk 2 above)
Comment (Keats) [reply content/blog/atomic-design-for-spas.md:36]: but you don't get the type checking, that's kinda useless

File: content/blog/atomic-design-for-spas.md (hunk 3 above)
Comment (Keats) [reply content/blog/atomic-design-for-spas.md]: Seen that but don't really agree with them, hence the "compelling"

File: content/blog/atomic-design-for-spas.md (hunk 4 above)
Comment (Keats) [reply content/blog/atomic-design-for-spas.md:60]: I don't use babel though :/

File: content/blog/atomic-design-for-spas.md (hunk 6 above)
Comment (Keats) [reply content/blog/atomic-design-for-spas.md]: I'll add pure to the list

File: content/blog/atomic-design-for-spas.md (hunk 5 above)
Comment (Keats) [reply content/blog/atomic-design-for-spas.md:100]: you get used to it!

File: content/blog/atomic-design-for-spas.md (hunk 7 above)
Comment (Keats) [reply content/blog/atomic-design-for-spas.md:139]: not sure what you mean by "HOC or container"

Comment (AdrienLemaire) [reply content/blog/atomic-design-for-spas.md:139]: Higher Order Component.  Opposite of SFC (Stateless Functional Component)

Maybe React jargon

File: content/blog/atomic-design-for-spas.md (hunk 2 above)
Comment (AdrienLemaire) [reply content/blog/atomic-design-for-spas.md:36]: you do ... with babel

Comment (Keats) [reply content/blog/atomic-design-for-spas.md:36]: Hmm no? Babel doesn't have the TS compiler builtin

File: content/blog/atomic-design-for-spas.md (hunk 7 above)
Comment (Keats) [reply content/blog/atomic-design-for-spas.md:139]: I know what it means, but why did you mention it there?

File: content/blog/atomic-design-for-spas.md (hunk 2 above)
Comment (AdrienLemaire) [reply content/blog/atomic-design-for-spas.md:36]: https://babeljs.io/docs/en/babel-preset-typescript

File: content/blog/atomic-design-for-spas.md (hunk 7 above)
Comment (AdrienLemaire) [reply content/blog/atomic-design-for-spas.md:139]: just in case you wanted to add more vocabulary ^^

File: content/blog/atomic-design-for-spas.md (hunk 2 above)
Comment (Keats) [reply content/blog/atomic-design-for-spas.md:36]: As the link says, it only parses TypeScript and doesn't do any typechecking. The babel plugin for TS is also limited and you can't use some TS features.

No mention of actual type checking anywhere.

**Comment by AdrienLemaire** [comment]
hmm ok, will need to be careful with that for our migration, thanks for the
heads up. I'll go smash the head of the guy who told me typescript
migration could be done in 1h :P

File: content/blog/atomic-design-for-spas.md [hunk 9]
```diff
@@ -0,0 +1,213 @@
++++
+title = "Atomic Design in Single Page Applications"
+description = "How I use Atomic Design to architect my SPAs in React"
+date = 2018-10-25
++++
+
+This post will outline how I architect visual part of the SPAs I work on. I used
```
Comment (deane) [inline content/blog/atomic-design-for-spas.md]: I feel like a transition is missing after the first sentence. Something like:
> This post will outline how I architect visual part of the SPAs I work on.

> But first, let me explain the issues I faced that led me there

File: content/blog/atomic-design-for-spas.md [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+`;
+
+export interface IData {
+  organizations: Array<IOrganization>;
+}
+
+export class AllOrgsQuery extends Query<IData> {}
+```
+
+You can find more about using TypeScript with Apollo [in their docs](https://www.apollographql.com/docs/react/recipes/static-typing.html).
```
Comment (deane) [inline content/blog/atomic-design-for-spas.md]: This whole section going through the `queries.tsx` file feels like it's not related to the article. 
Not sure what showing this code, and talking about typescript and graphql adds. And we haven't got to the atomic design part yet!

File: content/blog/atomic-design-for-spas.md [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+components which made things a bit tidier but was still very unsatisfying.
+Looking around for various ways to organise design elements, I found [Atomic Design](http://bradfrost.com/blog/post/atomic-web-design/).
+I am not going to dive deep into explaning what is Atomatic Design as the linked article does a very good job
+so I would recommend taking 2 minutes to read it beforehand.
+
+You can find a basic clone of [HackerNews](https://news.ycombinator.com/) on GitHub: <https://github.com/Keats/hn-react> made using
+the principles described in this post.
+
+Something missing from the repository that I use in my own projects is [StoryBook](https://storybook.js.org/) to easily have
+a visual overview of every component - an invaluable tool.
```
Comment (deane) [inline content/blog/atomic-design-for-spas.md]: I would push that reference to the end.

File: content/blog/atomic-design-for-spas.md (hunk 3 above)
Comment (deane) [reply content/blog/atomic-design-for-spas.md]: maybe instead of saying "Why do CSS in JS?" you should give an argument (or a few) for having a separate css file

File: content/blog/atomic-design-for-spas.md [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+Therefore molecules should be dumb components when possible but it isn't required.
+
+Here are some of the molecules from the same project:
+
+- `form` (`input` + `label`): using MobX for state but dumb otherwise
+- `dialog` (`overlay` + `button`): dumb
+- `toast-container` (`toast`): smart component that will display the currents `<Toast>` components from the store
+
+Whether `toast-container` belongs in the molecules folder is debatable
+but this is where it makes the most sense in my opinion.
```
Comment (deane) [inline content/blog/atomic-design-for-spas.md]: You could say why. By reading so far it seems like you're doing that because you want all the components in `atoms` to be dumb

File: content/blog/atomic-design-for-spas.md [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+
+If you need to use a molecule in another molecule, it probably means it should go in the organisms folder as we will now see.
+
+### Organisms
+
+As with molecules, there are two types of organisms I distinguish here:
+
+- components using one or more molecules
+- business-specific usage of molecule
```
Comment (deane) [inline content/blog/atomic-design-for-spas.md:154]: "business-specific" doesn't mean much here in my opinion. I would have used "a wrapper around a single molecule". Personal preference

File: content/blog/atomic-design-for-spas.md [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+
+Another good definition comes from the original article:
```
Comment (deane) [inline content/blog/atomic-design-for-spas.md]: Why another good definition? did you define it?

File: content/blog/atomic-design-for-spas.md [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+
+This post will outline how I architect visual part of the SPAs I work on. I used
+to put every components in a `components` folder, which obviously doesn't scale and make
+it hard finding a specific component. After that, I grouped at least the top level routing
+components which made things a bit tidier but was still very unsatisfying.
+Looking around for various ways to organise design elements, I found [Atomic Design](http://bradfrost.com/blog/post/atomic-web-design/).
+I am not going to dive deep into explaning what is Atomatic Design as the linked article does a very good job
+so I would recommend taking 2 minutes to read it beforehand.
+
+You can find a basic clone of [HackerNews](https://news.ycombinator.com/) on GitHub: <https://github.com/Keats/hn-react> made using
```
Comment (deane) [inline content/blog/atomic-design-for-spas.md:15]: Maybe say you're going to use it to illustrate the concepts/ideas in the post

File: content/blog/atomic-design-for-spas.md [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+- `sidebar`: uses only one molecule but is a distinct section of the interface
+- `login-form`: the `form` molecule set-up with the right fields and the GraphQL login query
+- `signup-form`: the `form` molecule set-up with the right fields and the GraphQL signup query
+
+As you can see, each form in the app gets its own component making testing/storybook very easy.
+Most of those forms components do not have style of their own as they just rely on the `form` molecule for that.
+
+### Templates
+
+My usage of templates is probably the biggest deviation from the original Atomic Design. If you look at most SaaS sites, you will notice
```
Comment (deane) [inline content/blog/atomic-design-for-spas.md]: Is it different? from my understanding, skimming the atomic design article, a template is a collection of organisms, with a layout. It seems to be exactly what you're doing

File: content/blog/atomic-design-for-spas.md (hunk 10 above)
Comment (Keats) [reply content/blog/atomic-design-for-spas.md]: Hmm it is to emphasize that component can be self-contained, I'll think a bit more about how to improve that
